metagraph/src/graph/alignment/alignment.hpp

```diff
@@ -23,41 +23,99 @@ namespace mtg {
+class Matching {
   public:
     typedef DeBruijnGraph::node_index node_index;
-    typedef DBGAlignerConfig::score_t score_t;
     typedef annot::binmat::BinaryMatrix::Column Column;
     typedef SmallVector<int64_t> Tuple;
     typedef Vector<Column> Columns;
     typedef Vector<Tuple> CoordinateSet;
+
+    Matching() {}
+
+    Matching(std::string_view query_view ,
+             std::vector<node_index>&& nodes,
+             bool orientation,
+             size_t offset,
+             size_t clipping,
+             size_t end_clipping)
+          : query_view_(query_view), nodes_(std::move(nodes)), orientation_(orientation),
+            offset_(offset), clipping_(clipping), end_clipping_(end_clipping) {}
+
+    virtual ~Matching() {}
+
+    size_t size() const { return nodes_.size(); }
+    bool empty() const { return nodes_.empty(); }
+    const std::vector<node_index>& get_nodes() const { return nodes_; }
+    std::string_view get_query_view() const { return query_view_; }
+    bool get_orientation() const { return orientation_; }
+    size_t get_offset() const { return offset_; }
+
+    virtual Cigar::LengthType get_clipping() const { return clipping_; }
+    virtual Cigar::LengthType get_end_clipping() const { return end_clipping_; }
+
+    virtual size_t trim_offset();
+
+    virtual size_t trim_query_prefix(size_t n,
+                                     size_t node_overlap,
+                                     const DBGAlignerConfig &config,
+                                     bool trim_excess_deletions = true);
+
+    virtual size_t trim_query_suffix(size_t n,
+                                     const DBGAlignerConfig &config,
+                                     bool trim_excess_deletions = true);
+
+    virtual size_t trim_reference_prefix(size_t n,
+                                         size_t node_overlap,
+                                         const DBGAlignerConfig &config,
+                                         bool trim_excess_insertions = true) {
+        return trim_query_prefix(n, node_overlap, config, trim_excess_insertions);
+    }
+
+    virtual size_t trim_reference_suffix(size_t n,
+                                         const DBGAlignerConfig &config,
+                                         bool trim_excess_insertions = true) {
+        return trim_query_suffix(n, config, trim_excess_insertions);
+    }
+
+    Columns label_columns;
+
+    // for each column in |label_columns|, store a vector of coordinate ranges
+    CoordinateSet label_coordinates;
+
+  protected:
+    std::string_view query_view_;
+    std::vector<node_index> nodes_;
+    bool orientation_;
+    size_t offset_;
+    Cigar::LengthType clipping_;
+    Cigar::LengthType end_clipping_;
+};
+
+// Note: this object stores pointers to the query sequence, so it is the user's
+//       responsibility to ensure that the query sequence is not destroyed when
+//       calling this class' methods
+class Alignment : public Matching {
```

Comment:
Good point, I wanted to get these changes working and forgot to structure these classes properly.

What I originally had in mind was that Matching would be an abstract parent class, and Seed and Alignment would inherit from it. Even though a Seed can be seen as a special case of Alignment, I don't want the Seed objects to store a bunch of information like Cigars and path spellings, since those are implicit.

Reply:
Sounds a bit too far-stretched. Maybe just make them all independent classes without any inheritance?

Reply:
It could be done that way, but then I'd have to duplicate the code for handling alignment orientation, annotations, path access, etc.

Reply:
So, I think it's better to remove this inheritance, even though some of their functions might be the same